File: internal/typesafe/typesafe.go
/*
 * Copyright (C) 2026 codesafe contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 * See COPYING in the project root for the full license.
 */
// typesafe 包是 TypeSafe System One HTTP API 的客户端：构造 questions、发送请求、解析按问题 id 返回的概率答案。
package typesafe

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"codesafe/internal/cache"
)

const endpoint = "https://api.typesafe.ai/v1/systemone"

// Client 调用 TypeSafe evaluation 端点。
type Client struct {
	apiKey  string
	model   string
	http    *http.Client
	retries int
	// Cache 开关响应缓存（bbolt，按请求 payload 的 SHA-256）；默认开。
	Cache bool
	// Refresh 强制跳过缓存重判（--refresh）。
	Refresh bool
}

// NewClient 返回使用给定 API key 的客户端，model 传空时使用 jev-latest。默认开缓存。
func NewClient(apiKey, model string) *Client {
	if model == "" {
		model = "jev-latest"
	}
	return &Client{
		apiKey:  apiKey,
		model:   model,
		http:    &http.Client{Timeout: 120 * time.Second},
		retries: 4,
		Cache:   true,
	}
}

// Question 是单个问题（noul 或 choice）。
type Question struct {
	Type         string            `json:"type"`
	Instructions string            `json:"instructions"`
	Criteria     map[string]string `json:"criteria,omitempty"`
}

// Noul 构造一个 yes/no 问题，criteria 描述 true/false 各自的含义。
func Noul(instructions string, criteria map[string]string) Question {
	return Question{Type: "noul", Instructions: instructions, Criteria: criteria}
}

// Choice 构造一个选项问题，criteria 为 选项名 -> 说明。
func Choice(instructions string, criteria map[string]string) Question {
	return Question{Type: "choice", Instructions: instructions, Criteria: criteria}
}

// request 是 POST /v1/systemone 的请求体。
type request struct {
	State     any                 `json:"state"`
	Model     string              `json:"model"`
	Questions map[string]Question `json:"questions"`
}

// Response 是 evaluation 端点的响应；每个问题的答案按其 id 键控。
type Response struct {
	Model   string                     `json:"model"`
	Answers map[string]json.RawMessage `json:"answers"`
	Usage   Usage                      `json:"usage"`
}

// Usage 是一次请求的 token 用量。
type Usage struct {
	InputTokens  int `json:"input_tokens"`
	OutputTokens int `json:"output_tokens"`
}

// apiError 是端点返回的错误体。
type apiError struct {
	Error struct {
		Message string `json:"message"`
	} `json:"error"`
}

// rateLimitError 表示 429，携带 Retry-After 秒数（可为 0）。
type rateLimitError struct{ retryAfter time.Duration }

func (e *rateLimitError) Error() string { return "rate limited" }

// Answer 是一个问题的原始答案（按类型字段填充）。
type Answer struct {
	// noul
	Noul float64
	// choice
	Choice        string
	Probabilities map[string]float64
	Confidence    float64
}

// Evaluate 对同一 state 求值所有 questions，返回每个问题 id 的 Answer 及 token 用量。
func (c *Client) Evaluate(ctx context.Context, state any, questions map[string]Question) (map[string]Answer, Usage, error) {
	body := request{State: state, Model: c.model, Questions: questions}
	payload, err := json.Marshal(body)
	if err != nil {
		return nil, Usage{}, err
	}
	// 缓存键 = 真实发往上游的请求字节 SHA-256；命中且未过期则直接复用。
	key := cache.Key(payload)
	if c.Cache && !c.Refresh {
		if e, ok := cache.Get(key); ok {
			var out Response
			if json.Unmarshal(e.Answers, &out.Answers) == nil {
				if ans, err := decodeAnswers(out.Answers); err == nil {
					var u Usage
					json.Unmarshal(e.Usage, &u)
					return ans, u, nil
				}
			}
		}
	}
	var lastErr error
	for attempt := 0; attempt <= c.retries; attempt++ {
		if attempt > 0 {
			if !sleep(ctx, backoff(attempt, lastErr)) {
				return nil, Usage{}, ctx.Err()
			}
		}
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(payload))
		if err != nil {
			return nil, Usage{}, err
		}
		req.Header.Set("Authorization", "Bearer "+c.apiKey)
		req.Header.Set("Content-Type", "application/json")
		resp, err := c.http.Do(req)
		if err != nil {
			lastErr = err
			continue
		}
		raw, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			lastErr = err
			continue
		}

		if resp.StatusCode == http.StatusOK {
			var out Response
			if err := json.Unmarshal(raw, &out); err != nil {
				return nil, Usage{}, fmt.Errorf("解析响应失败: %w", err)
			}
			ans, err := decodeAnswers(out.Answers)
			if err != nil {
				return nil, Usage{}, err
			}
			if c.Cache {
				rawAns, _ := json.Marshal(out.Answers)
				rawUse, _ := json.Marshal(out.Usage)
				cache.Set(key, cache.Entry{Answers: rawAns, Usage: rawUse})
			}
			return ans, out.Usage, nil
		}
		if resp.StatusCode == http.StatusTooManyRequests || resp.StatusCode >= 500 {
			lastErr = retryable(resp, raw)
			continue
		}
		return nil, Usage{}, fatal(resp, raw)
	}
	return nil, Usage{}, fmt.Errorf("请求失败（重试 %d 次后）: %w", c.retries, lastErr)
}

// decodeAnswers 把 answers map 里每个答案按字段解出（noul 概率、choice 选中项/概率/置信度）。
func decodeAnswers(answers map[string]json.RawMessage) (map[string]Answer, error) {
	out := make(map[string]Answer, len(answers))
	for id, raw := range answers {
		var a struct {
			Noul          *float64           `json:"noul"`
			Choice        string             `json:"choice"`
			Probabilities map[string]float64 `json:"probabilities"`
			Confidence    float64            `json:"confidence"`
		}
		if err := json.Unmarshal(raw, &a); err != nil {
			return nil, fmt.Errorf("解析问题 %q 的答案失败: %w", id, err)
		}
		an := Answer{Choice: a.Choice, Probabilities: a.Probabilities, Confidence: a.Confidence}
		if a.Noul != nil {
			an.Noul = *a.Noul
		}
		out[id] = an
	}
	return out, nil
}

// retryable 把可重试的响应转成 error，429 时解析 Retry-After。
func retryable(resp *http.Response, raw []byte) error {
	if resp.StatusCode == http.StatusTooManyRequests {
		d := 0.0
		if s := resp.Header.Get("Retry-After"); s != "" {
			fmt.Sscanf(s, "%f", &d)
		}
		return &rateLimitError{retryAfter: time.Duration(d * float64(time.Second))}
	}
	return fmt.Errorf("上游错误 %d: %s", resp.StatusCode, truncate(raw))
}

// fatal 把不可重试的响应转成带上游消息的错误。
func fatal(resp *http.Response, raw []byte) error {
	var ae apiError
	if json.Unmarshal(raw, &ae) == nil && ae.Error.Message != "" {
		return fmt.Errorf("API 错误 %d: %s", resp.StatusCode, ae.Error.Message)
	}
	return fmt.Errorf("API 错误 %d: %s", resp.StatusCode, truncate(raw))
}

// backoff 计算第 n 次重试的等待时长；429 优先用 Retry-After。
func backoff(attempt int, lastErr error) time.Duration {
	var rl *rateLimitError
	if asRateLimit(lastErr, &rl) && rl.retryAfter > 0 {
		return rl.retryAfter
	}
	return time.Duration(attempt*attempt) * time.Second
}

// asRateLimit 判断 err 是否为 rateLimitError。
func asRateLimit(err error, out **rateLimitError) bool {
	for err != nil {
		if rl, ok := err.(*rateLimitError); ok {
			*out = rl
			return true
		}
		return false
	}
	return false
}

// sleep 等待 d 或 ctx 取消；返回 false 表示 ctx 已取消。
func sleep(ctx context.Context, d time.Duration) bool {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-t.C:
		return true
	}
}

// truncate 截断错误响应体用于日志。
func truncate(b []byte) string {
	s := strings.TrimSpace(string(b))
	if len(s) > 300 {
		return s[:300] + "…"
	}
	return s
}

File: extensions/pi/core/typesafe.ts
/*
 * Copyright (C) 2026 codesafe contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 * See COPYING in the project root for the full license.
 */
// typesafe.ts — TypeSafe System One HTTP client: build questions, POST to the
// evaluate endpoint, decode per-question answers (noul probability / choice).

export const ENDPOINT = "https://api.typesafe.ai/v1/systemone";
export const DEFAULT_MODEL = "jev-latest";

/** A single System One question (noul yes/no, or choice over a label set). */
export interface Question {
	type: "noul" | "choice";
	instructions: string;
	criteria: Record<string, string>;
}

/** Build a yes/no question; criteria maps "true"/"false" to their meanings. */
export function noul(instructions: string, criteria: Record<string, string>): Question {
	return { type: "noul", instructions, criteria };
}

/** Build a choice question; criteria maps option name -> description. */
export function choice(instructions: string, criteria: Record<string, string>): Question {
	return { type: "choice", instructions, criteria };
}

/** Raw decoded answer for one question id. */
export interface Answer {
	/** noul probability that the answer is "true". */
	noul: number;
	/** choice-selected label. */
	choice: string;
	/** per-option probabilities. */
	probabilities: Record<string, number>;
	/** model-reported confidence. */
	confidence: number;
}

/** Token usage for one request. */
export interface Usage {
	input_tokens: number;
	output_tokens: number;
}

/** Error raised when the upstream reports a context/token limit. */
export class TokenLimitError extends Error {}

/** Thrown for non-retryable upstream errors, carrying the upstream message. */
export class ApiError extends Error {
	constructor(
		message: string,
		readonly status: number,
	) {
		super(message);
	}
}

interface ApiResponse {
	answers?: Record<string, unknown>;
	usage?: Usage;
	error?: { message?: string };
}

function isTokenLimitMessage(msg: string): boolean {
	return /max_tokens|tokens_exceeded|context/i.test(msg);
}

function sleep(ms: number): Promise<void> {
	const { promise, resolve } = Promise.withResolvers<void>();
	setTimeout(resolve, ms);
	return promise;
}

/** Client for the TypeSafe evaluate endpoint, with retry + in-memory cache. */
export class Client {
	private readonly apiKey: string;
	private readonly model: string;
	private readonly retries = 3;
	/** Simple in-process response cache keyed by serialized request bytes. */
	private readonly cache = new Map<string, { answers: Record<string, Answer>; usage: Usage }>();
	/** When true, read-through cache is enabled. */
	cacheOn = true;
	/** When true, bypass reads (but still writes) — refresh semantics. */
	refresh = false;

	constructor(apiKey: string, model = DEFAULT_MODEL) {
		this.apiKey = apiKey;
		this.model = model || DEFAULT_MODEL;
	}

	/** Evaluate all questions against the same state in one request. */
	async evaluate(
		state: unknown,
		questions: Record<string, Question>,
	): Promise<{ answers: Record<string, Answer>; usage: Usage }> {
		const body = { state, model: this.model, questions };
		const key = JSON.stringify(body);
		if (this.cacheOn && !this.refresh) {
			const hit = this.cache.get(key);
			if (hit) return { answers: hit.answers, usage: hit.usage };
		}
		const payload = key;

		let lastErr: unknown;
		for (let attempt = 0; attempt <= this.retries; attempt++) {
			if (attempt > 0) {
				const is429 = lastErr instanceof ApiError && lastErr.status === 429;
				await sleep(is429 ? Math.min(2000 * attempt, 8000) : Math.min(500 * 2 ** attempt, 4000));
			}
			let resp: Response;
			try {
				resp = await fetch(ENDPOINT, {
					method: "POST",
					headers: {
						Authorization: `Bearer ${this.apiKey}`,
						"Content-Type": "application/json",
					},
					body: payload,
				});
			} catch (err) {
				lastErr = err;
				continue;
			}
			const raw = await resp.text();
			if (resp.status === 200) {
				let out: ApiResponse;
				try {
					out = JSON.parse(raw) as ApiResponse;
				} catch (err) {
					throw new ApiError(`failed to parse response: ${String(err)}`, resp.status);
				}
				const answers = decodeAnswers(out.answers ?? {});
				const usage = out.usage ?? { input_tokens: 0, output_tokens: 0 };
				if (this.cacheOn) this.cache.set(key, { answers, usage });
				return { answers, usage };
			}
			// Retryable: 429 + 5xx.
			if (resp.status === 429 || resp.status >= 500) {
				lastErr = new ApiError(`HTTP ${resp.status}: ${truncate(raw)}`, resp.status);
				continue;
			}
			// Fatal — surface upstream message; classify token-limit for callers.
			const msg = parseErrorMessage(raw) || `HTTP ${resp.status}`;
			if (isTokenLimitMessage(msg)) throw new TokenLimitError(msg);
			throw new ApiError(msg, resp.status);
		}
		throw new ApiError(`request failed after ${this.retries} retries: ${String(lastErr)}`, 0);
	}
}

/** Decode the answers map into typed Answer objects per question id. */
function decodeAnswers(raw: Record<string, unknown>): Record<string, Answer> {
	const out: Record<string, Answer> = {};
	for (const [id, v] of Object.entries(raw)) {
		const r = (v ?? {}) as Record<string, unknown>;
		out[id] = {
			noul: num(r.noul ?? r.probability ?? r.p),
			choice: str(r.choice ?? r.selected ?? r.answer),
			probabilities: (r.probabilities as Record<string, number>) ?? {},
			confidence: num(r.confidence),
		};
	}
	return out;
}

function num(v: unknown): number {
	return typeof v === "number" ? v : 0;
}
function str(v: unknown): string {
	return typeof v === "string" ? v : "";
}
function parseErrorMessage(raw: string): string {
	try {
		const o = JSON.parse(raw) as ApiResponse;
		return o.error?.message ?? "";
	} catch {
		return truncate(raw);
	}
}
function truncate(s: string): string {
	return s.length > 300 ? `${s.slice(0, 300)}…` : s;
}
